docs/lexer.md

```diff
@@ -0,0 +1,51 @@
+# Design
+
+- State machine based lexer
+
+## Additional reading
+
+Concept: Lexer receivs a CharProvider, which holds a Reader and will lazily read from STDIN.  
+This allows the lexer to ignore or consume newlines without producing a newline tokens, namely in any state other than the EMPTY state.
+
+## Recursive lexing (i.e. expansions)
+
+Some states expect to return to the previous state:
+- Backslash
+- CommandSubstitution
+- DoubleQuotes
+- SingleQuotes
+- ParameterExpansion
+- ArithmeticExpansion
+
+This means the precondition to enter these states, is to push the current state onto the stack, and enter those states.
```

Comment:
Do we have an assert in these states to make sure that the stack of states is not empty?

Reply:
No, good point. I'll add them

Reply:
Maybe we should add this logic to the `getStateHandler` or `nextToken` function, so we can both assert that when a non-root state is called, the stack isn't empty, and also vice versa, when a root state is called, the stack should be empty

Reply:
I think it's fine to do it inside of the handlers, keeps it simpler imo. Maybe if there will be many more states in the future but I don't imagine that will be the case.